Update wrappers documentation

LetteraUnica opened the pull request:
# Description
* Made clear, through a usage example, to reset a wrapped environment before calling any other methods.
* Specified in the basic usage documentation that an environment should be resetted before use.

Fixes #877

## Type of change

- Documentation change

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my change

elliottower commented:
Got a chance to test your issue on my machine and got the same error. 

I see you did what you described in the issue and copied the code from the base AEC environment defined in env.py: https://github.com/Farama-Foundation/PettingZoo/blob/master/pettingzoo/utils/env.py#L184, and changed self._cumulative_rewards to self.env._cumulative_rewards. If it solves the problem that's good, but I'm not sure it's the cleanest way to do it, it seems messy to have self.env._ in one line and then self.terminations in the next line. I imagine this type of issue is handled in the parallel env base wrappers as well as gymnasium most likely, I'll look into that and see what I  can find. Maybe @jjshoots or @pseudo-rnd-thoughts can chime in as they're much more familiar with these details than I am.

As a side comment, it appears the documentation doesn't even list `last()`, ([AEC](https://pettingzoo.farama.org/api/aec/) and [parallel](https://pettingzoo.farama.org/api/parallel/)) despite it being front and center of the basic usage example used widely in other examples. There should probably be documentation showing how to use parallel envs as well as AEC envs (and ideally tutorials for creating both).

pseudo-rnd-thoughts commented:
There are two solutions, either raise a more helpful error (`NotImplementedError`) in `BaseWrapper.last` as I think that `ParallelEnv` don't use it. 
Or implement the function using the base environment's `self.env.last()` and saving the data

elliottower commented:
> There are two solutions, either raise a more helpful error (`NotImplementedError`) in `BaseWrapper.last` as I think that `ParallelEnv` don't use it. Or implement the function using the base environment's `self.env.last()` and saving the data
> 
> Also could we have a test to check that this fixes the error with the expected behaviour

Any chance you could look into this? @LetteraUnica If not I can make a separatePR with Mark’s suggested changes

LetteraUnica commented:
I'm looking into it right now

elliottower commented:
> I'm looking into it right now

Awesome, appreciate the help.

LetteraUnica commented:
I explored the ways you proposed to solve this:

* Raising a `NotImplementedError`. However if a user only wants to change the observation, for example rescale the output of `env.observe()`, then the user must also override the `.last()` method, which is inconvenient
* Implementing the function using the base environment's `self.env.last()` and saving the data. If I understand correclty this only takes one line of code.
  ```python
  def last(self, observe: bool = True):
        return self.env.last(observe=observe) 
  ```
  However using `env.last()` can introduce bugs, as in the previous point, suppose the user wants to scale the observation and thus overrides the `.observe()` method, then `.observe()` would return the correctly scaled observation, while `.last()` will call `env.last()` which returns the original non-scaled observation.

I have also come up with another way to fix the problem:
* Since `_cumulative_rewards` can't be accessed even by the `BaseWrapper` class, we could make it a property:
    ```python
    @property
    def _cumulative_rewards(self):
        return self.env._cumulative_rewards
    ```
This way, `_cumulative_rewards` can be accessed and its value will always be the env' value, furthermore we don't need to override `.last()` in `BaseWrapper`.

Tell me what you would like to implement, as I don't know how each change would impact other parts of the library.

elliottower commented:
The property approach sounds most reasonable to me but I’m not sure if it’ll introduce any other issues, have you run pytests? Could also check how gymnasium does it or if there are or their similar uses of @property. I think @jjshoots @pseudo-rnd-thoughts will have more input.

pseudo-rnd-thoughts commented:
I guess it makes sense if wrappers don't modify `_cumulative_rewards` but this more just shows issues in the API imho

LetteraUnica commented:
> The property approach sounds most reasonable to me but I’m not sure if it’ll introduce any other issues, have you run pytests? Could also check how gymnasium does it or if there are or their similar uses of @Property. I think @jjshoots @pseudo-rnd-thoughts will have more input.

I managed to run pre-commit checks, however I encountered an error when trying to execute `pytest -v`. The problem is GLIBCXX_3.4.29 not found, I tried the suggestions from stackoverflow, however it still doesn't work.

elliottower commented:
Hey @LetteraUnica thanks again for the work, can you sync your branch with master? There was an issue causing CI to fail but it’s been fixed.

LetteraUnica commented:
Hi @elliottower I synched my branch with master and added more tests, in particular now I test both the Go and Chess environments, however it can also be extended to other environments.

I also renamed the file as you requested, however all the tests in the `pettingzoo/test` folder follow the naming convention `"name"+"_test"`, that is why I called my test `wrapper_test`.

I'd like to ask you if there is already a function in the repository that checks if two calls of the `.last()` method return the same result, like I did in the `check_equal` function, so that we avoid code duplication.

elliottower commented:
> Hi @elliottower I synched my branch with master and added more tests, in particular now I test both the Go and Chess environments, however it can also be extended to other environments.
> 
> I also renamed the file as you requested, however all the tests in the `pettingzoo/test` folder follow the naming convention `"name"+"_test"`, that is why I called my test `wrapper_test`.
> 
> I'd like to ask you if there is already a function in the repository that checks if two calls of the `.last()` method return the same result, like I did in the `check_equal` function, so that we avoid code duplication.

Great, thanks for all the hard work on this. And yep you're right, I was mistaken about the naming convention. Good catch. But I'm not sure these tests are actually necessary, if you look at some of the other tests like https://github.com/Farama-Foundation/PettingZoo/blob/master/test/all_parameter_combs_test.py#L228, they run the API test which includes a `play_test()` which checks accumulated rewards. I just approved the workflow so we'll see if the existing tests pass, but my thought is that this shouldn't change the functionality of anything, just prevents the edge case error you experienced. 

Could you make an example script which runs the exact case you had in your original post and tests to see if that case throws the error?

LetteraUnica commented on test/test_wrapper.py:
This test is the problem I had before with the library, I couldn't call the .last method of a wrapped environment as it tried to access _cumulative_rewards and then throw an error.

elliottower replied:
I feel like you could change this test to just take a regular environment like chess and test wrapping it and then calling last. It would probably make sense to be added to the API test as that’s a pretty core API functionality. The custom environment code is a little messy and overall it doesn’t fit very well within the existing tests imo.

jjshoots replied:
@LetteraUnica Is there a possibility of integrating this test with one of the other tests, such as the unwrapped tests? Just a curiousity. It doesn't feel like it's necessary to spin up a whole new file to test one functionality, but I could be wrong of course.

LetteraUnica commented on pettingzoo/.idea/inspectionProfiles/profiles_settings.xml:
Removed, I don't know how it managed to get there

LetteraUnica commented:
I noticed that the original bug was caused by not resetting the wrapped environment before use, even though I called `env.reset()` on the base environment. Talking with @elliottower about the problem, we think that the best option would be to update the documentation, making it clear to reset a wrapped environment before use. So I edited the title and the body of the PR to reflect this change.

jjshoots reviewed (approved):
This looks much more sensible now, thanks for the contribution!
